src/platform/android/java/chip/platform/NsdManagerServiceResolver.java

```diff
@@ -223,4 +255,47 @@ public void removeServices() {
     registrationListeners.clear();
     mMFServiceName.clear();
   }
+
+  /**
+   * The Android NsdManager calls back on the NsdManager.ResolveListener with a
+   * FAILURE_ALREADY_ACTIVE(3) if any application code calls resolveService() on it while the
+   * resolve operation is already active (from another call made previously). An object of
+   * NsdManagerResolverAvailState allows NsdManagerServiceResolver to synchronize on the usage of
+   * NsdManager's resolveService() API
+   */
+  public static class NsdManagerResolverAvailState {
+    private static final String TAG = NsdManagerResolverAvailState.class.getSimpleName();
+
+    private Lock lock = new ReentrantLock();
+    private Condition condition = lock.newCondition();
+    private boolean busy = false;
+
+    /**
+     * Waits if the NsdManager is already busy with resolving a service. Otherwise, it marks it as
+     * busy and returns
+     */
+    public void markBusyIfFreeOrWait() {
```

Comment:
nit: method names with `or` are generally harder to reason about because they include separate action paths. How about naming this something singular, like `acquireBusy` or `lockBusy` or similar?

Reply:
Done. Called it acquireResolver()